dusk/grammar.py

```diff
@@ -554,7 +557,13 @@ def field_index(self, field: DuskField, vindex=None, hindex=None):
                 f"Invalid horizontal offset for field '{field.sir.name}'!"
             )
 
-        return make_unstructured_offset(True), voffset, vbase
+        if not self.ctx.location.is_dense(field_dimension):
+            raise DuskSyntaxError(
+                f"Invalid horizontal offset for field '{field.sir.name}'!"
+                "sparse fields do not take an offset (only non-offset read makes sense)"
+            )
```

Comment:
I think this should be the very first check. Even before:
https://github.com/dawn-ico/dusk/blob/a6d9fa7f977b26f09f9948ee1956d09247850fe3/dusk/grammar.py#L510-L516

Then we'll need (+ move initialization of `field_dimension` further up):
```suggestion
        if hindex is not None and not self.ctx.location.is_dense(field_dimension):
            raise DuskSyntaxError(
                f"Invalid horizontal offset for field '{field.sir.name}',"
                "sparse fields do not take an offset (only non-offset read makes sense)!"
            )
```

Yes, this whole code is way too complicated (probably the most complicated part in all of dusk). We should probably look into simplifying this at some point. But then the whole _location semantics_ are pretty complicated as well.

In it's current form I think it will allow:
```python
@stencil
def sparse_field_with_hoffset(a: Field[Edge], b: Field[Edge > Cell > Edge]):
  with levels_upward:
    # horizontal index for sparse field
    a = reduce_over(Edge > Cell > Edge, b[Edge])
```
But I think we want to disallow horizontal indices for all sparse fields in general (independent of the context).

Reply:
good catch on the unchecked case!

moved check up as far as possible according to discussion on slack